// ionic/commands/enterprise/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_framework_1 = require("@ionic/cli-framework");
const namespace_1 = require("../../lib/namespace");
class EnterpriseNamespace extends namespace_1.Namespace {
    getMetadata() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                name: 'enterprise',
                summary: 'Manage Ionic Enterprise features',
                description: `
Commands to help manage Ionic Enterprise[^enterprise-edition] subscriptions.
      `,
                footnotes: [
                    {
                        id: 'enterprise-edition',
                        url: 'https://ionicframework.com/enterprise-edition',
                        shortUrl: 'https://ion.link/enterprise',
                    },
                ],
                groups: [cli_framework_1.NamespaceGroup.Paid],
            };
        });
    }
    getCommands() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new namespace_1.CommandMap([
                ['register', () => tslib_1.__awaiter(this, void 0, void 0, function* () { const { RegisterCommand } = yield Promise.resolve().then(() => require('./register')); return new RegisterCommand(this); })],
            ]);
        });
    }
}
exports.EnterpriseNamespace = EnterpriseNamespace;
